designs/disruption-controls.md

```diff
@@ -0,0 +1,118 @@
+
+These controls will include a new `Disruption` block in the `v1beta1` `NodePool` that will contain all scale-down behavioral fields. Initially, this will include `Budgets`, which define a (1) parallelism of how many nodes can be deprovisioned at a time and a (2) cron schedule that determines when the parallelism applies, and a `ConsolidateAfter` field that will allow users to affect the speed at which Karpenter will scale down underutilized nodes. Future scale-down behavioral fields should be colocated with these fields.
+
+## Proposed Spec
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized || WhenEmpty
+    consolidateAfter: 10m || Never # metav1.Duration
+    expireAfter: 10m || Never # Equivalent to v1alpha5 TTLSecondsUntilExpired
+    budgets:
+    # On Weekdays during business hours, don't do any deprovisioning.
+    - crontab: "0 9 * * mon-fri"
+      duration: 8h
+      maxUnavailable: 0
+    # Every other time, only allow 10 nodes to be deprovisioned simultaneously
+    - maxUnavailable: 10
+```
+
+## Code Definition
+
+```go
+type Disruption struct {
+    {...}
+    // Budgets is a list of Budgets.
+    // If there are multiple active budgets, the most restrictive budget's maxUnavailable is respected.
+    Budgets []Budget `json:"budgets,omitempty" hash:"ignore"`
+    // ConsolidateAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Consolidation.
+    ConsolidateAfter *NillableDuration `json:"consolidateAfter" hash:"ignore"`
+    // ConsolidationPolicy determines how Karpenter will consider nodes
+    // as candidates for Consolidation.
+    // WhenEmpty uses the same behavior as v1alpha5 TTLSecondsAfterEmpty
+    // WhenUnderutilized uses the same behavior as v1alpha5.ConsolidationEnabled: true
+    ConsolidationPolicy string `json:"consolidationPolicy" hash:"ignore"`
+}
+// Budget specifies periods of times where Karpenter will restrict the
+// number of Node Claims that can be terminated at a time.
+// Unless specified, a budget is always active.
+type Budget struct {
+    // MaxUnavailable dictates how many Node Claims owned by this NodePool
+    // can be terminating at once.
+    // This only respects and considers nodes with a DeletionTimestamp set.
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable" hash:"ignore"`
```

Comment:
Wonder if making this resource based instead of node bases makes more sense.

With karpenter nodes having varied resources treating nodes as the same seems potentially risky/counter intuitive. 

I wondering about making this field a % based only which would be a % of the NodePool's current resources.

Would be slightly more difficult logic, but it seems to match karpenters patterning of having different nodes with different compute/resourcing.

Reply:
Great questions. I'll try to tackle them one at a time. 

> Wonder if making this resource based instead of node bases makes more sense.

We've definitely considered it, and this can definitely be done as a follow-up in the future. Ultimately Karpenter considers disrupting nodes on a per-node basis, where currently most disruption mechanisms (expiration, drift, single consolidation) will only disrupt one node at a time. Since Karpenter thinks about this in terms of number of nodes, and not the capacity, it always allows Karpenter to make disruption progress regardless of the instance sizes. 

Using resource-based disruption budgets can get users into funky situations where they might define a hard CPU value (10cpu), where all their instances in the cluster have at least 16cpu of capacity. For application developers who only care if their applications are disrupted or not, they are only guaranteed no disruptions if they have `maxUnavailable` set to 0, opting out all instance types.

> I wondering about making this field a % based only which would be a % of the NodePool's current resources.

For a set of nodes with varied instance sizes, using a percent shouldn't be used to drive a minimum percentage of application disruptions. This is meant to be supplemental to PDBs, where users define the level of downtime their applications can tolerate. It's important to note that this won't be incurring any more downtime than PDBs/do-not-evict/ other blocking mechanisms would allow.

Reply:
@njtran  how would you feel about maxUnavailable pertaining to pods rather than nodes? 
The peculiar thing about using nodes here is they're theoretically different sizes aren't they? So I might need to replace ones that have drifted but move more than the majority of my workload because the first N nodes I replace happen to be large ones as opposed to small ones? I think pod percenatge would slow this down. 

May be off base here though.

Reply:
You should use the Pod Disruption Budgets to control the level of `maxUnavailable` that you want to enforce for your applications. You won't need to manually replace any of your pods/nodes as that's all programmatically done by Karpenter.

Reply:
@njtran  Already in the Pod Disruption Budgets fanclub but most clusters only have a few for their most imporant applications. There is still often an expectation of cluster operators that they move workloads to new nodes gradually (may be jobs or other transient workload) . Eventually you will also get an ask to pause between chunks of movement to give alerts time to detect funniness. 

Number or percent of Nodes was a good proxy for this in the past when usually pools of nodes were fairly homogenous but with karpenter they can be quite different and there's not a great way of saying replacing these 3 HUGE machines causes much more churn than 

My hope is that pod count is a a good replacement proxy in the karpenter world. if you move 5% of pods at random even without the dilligence of having everything protecte with PDBs you stand a good chance of letting new pods come up and detecting issues when they exist. 

My 2 cents hope that was more clear.

 - Miller

Reply:
👍 on what @paulgmiller is describing. 

On a binary level, a lot of alerts have threshold time periods before they'll fire so many times we like making any changes at a controlled rate to ensure anything unexpected has time to hit those thresholds. 

As far as the ask for pod count, as a cluster operator I do not understand (and my users don't really either) all of the interdependencies of their applications. While a PDB can work for the health of an individual deployment, something like controlling total churn in the cluster can help gracefully degrade instead of outright breaking an application when one of these interdependencies causes issues. 

I'd also point out things like kube-proxy where if you have a sudden, large change you can cause issues with cluster stability because of the time it takes for reconciliation (say, 2 large nodes with 100+ pods each vs 2 small nodes with 30 pods each).

Reply:
Thanks for the feedback. 

You could imagine extending this to custom resources, ephemeral storage, pods, cpu, memory, or whatever else is used for bin-packing in the cluster.

If we're able to zero the chances that users can get themselves in tough situations with a budget and unfortunate instance type selections, Karpenter will always be able to converge towards an optimal cluster state to save on cost, which is the biggest pro in my mind. 

I'm 100% open to explore expanding this to more resource types in the future, but want to keep the initial implementation as simple as possible, leaving it open for expansion later.